hubble-relay: add support for (m)TLS

rolinh opened the pull request:
This PR adds support for TLS and mTLS to Hubble Relay. This allows to configure (m)TLS for the server as well as when establishing connections to peers in the cluster.

Note: This PR depends on #12906.

Ref: #11224

rolinh committed:
hubble/relay: add possibility to provide a hostname to ClientConn

Update the ClientConnBuilder interface such that the ClientConn function
takes hostname as an additional parameter. This new parameter can be
ignored if not needed but is useful when establishing a secure
connection to verify a certificate's hostname.

As the ClientConnBuilder interface is updated, adapt existing
implementations for the change. The GRPCClientConnBuilder implementation
may not be provided with TLS configuration, in which case connection is
secured with TLS and the certificate is verified against the provided
hostname.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: add options to configure transport security

Add a set of options to configure transport security for the relay
server and for the connection to peers. It is currently disabled for
Hubble Relay.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble-relay: add viper to support config via ENV variables

Introduce viper so that options can be set via environment variables.
Using flags is still supported and flags have priority over environment
variables.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
helm: create a ConfigMap for Hubble Relay and mount it as config

Create a configmap to set config options for Hubble Relay. This avoids
passing flags via args. Support for reading from a config needs to be
added to Hubble Relay.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble-relay: support reading config from a file

This allows mounting Hubble Relay's ConfigMap as YAML config file.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: split WithInsecure option for server and client

One needs to be able to allow Hubble Relay to establish its connections
to peers in the cluster over plaintext (`WithInsecureClient()`) and/or
to server over plaintext (`WithInsecureServer()`). Update the
`WithInsecure()` option to distinguish the two cases.

While here, rename internal variables in order to prevent confusion
between the two.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble-relay: add option to set (m)TLS with Hubble server instances

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble-relay: add options to set TLS for the server

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
helm: generate separate TLS certs for Hubble Relay's client and server

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
helm: enable mTLS for Hubble Relay

Enable mTLS between Hubble Relay and Hubble server instances but disable
TLS by default for Hubble Relay as this would mean that clients need to
add the Hubble CA to trust Hubble Relay's certificate.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/peer: add TLS information to the Peer type

This allows to know whether TLS should be used when connecting to the
peer's server endpoint.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: include TLS info about peers

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/relay: pass TLS server name to peer conn builder

If TLS is enabled, this allows to connect via (m)TLS.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
helm: set Hubble Relay service port dynanically

When the Hubble Relay service port is not explicitly provided, set it to
port 443 if TLS is enabled and to port 80 otherwise.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
helm: update experimental-install.yaml for recent Hubble Relay changes

Flags were moved to a configmap, thus re-generate the file.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

kkourt reviewed (approved):
LGTM.

gandro commented on pkg/hubble/relay/server/option.go:
```diff
@@ -212,3 +212,13 @@ func WithClientMTLSFromCert(ca *x509.CertPool, cert tls.Certificate) Option {
+// WithInsecureClient disables transport security for connection to Hubble
+// server instances. Transport security is required to WithInsecureClient is
+// set (not recommended).
+func WithInsecureClient() Option {
```
nit: I find the client/server terminology a bit ambiguous here. Yes, hubble-relay acts as a client, but it also has clients such as UI/CLI that connect to it.

Can we call the client-tls `peer-tls` instead, i.e. name this `WithInsecurePeer` and the options `tls-peer-cert-file` and so forth?

rolinh replied:
That's a valid point. I'll try to come up with a terminology that is consistent and not confusing and will amend the PR.

rolinh replied:
@kAworu opened #13094

gandro commented on hubble-relay/cmd/serve/serve.go:
```diff
@@ -107,11 +111,26 @@ func New(vp *viper.Viper) *cobra.Command {
+	flags.Bool(
+		keyTLSServerDisabled,
+		false,
+		"Disable (m)TLS for the server and allow clients to connect over plaintext.",
```
```suggestion
		"Disable TLS for the server and allow clients to connect over plaintext.",
```

gandro commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -319,8 +319,7 @@ func (m *PeerManager) connect(p *peer, ignoreBackoff bool) {
 	}).Debugf("Connecting peer %s...", p.Name)
-	//FIXME: provide hostname to ClientConn
-	conn, err := m.opts.clientConnBuilder.ClientConn(p.Address.String(), "")
+	conn, err := m.opts.clientConnBuilder.ClientConn(p.Address.String(), p.TLSServerName)
```
Should we throw a warning if `p.TLSEnabled=true` but hubble-relay was not configured with any certs for TLS for peers?

rolinh replied:
It would be nice to throw a warning but in practice it's a little difficult as the pool manager is not aware whether relay is configured with certs for TLS or not (the `ClientConnBuilder` is set when instantiating a new `Server` and at this point the `Server` is not aware if peers are TLS enabled or not and the pool manager is not aware of how `ClientConn()` establishes the connection (it's an interface)).

TLDR: it would require quite some changes to implement. If you feel it's a must have, I suggest to open an issue and address this later.

kaworu replied:
We can check `m.opts.clientConnBuilder.TLSConfig` right? Alternatively, `ClientConn()` could return an error when it has a TLS config while called with an empty server name.

rolinh replied:
> We can check m.opts.clientConnBuilder.TLSConfig right?

I'm not really sure what you are trying to imply here.  The current code uses the `GRPCClientConnBuilder` as the implementation of the `ClientConnBuilder` interface. This implementation opts to connect using TLS if a TLS config is provided and uses plaintext otherwise.

I think that what @gandro is pointing out is that the peer service already reports whether a peer is configured with TLS or not (see `Peer.TLSEnabled`) but this information is discarded by the pool manager which, by abstraction, knows nothing about how the connection to peers is established. In order to change this, the `ClientConnBuilder` interface should be modified such that the caller defines whether TLS is required or not.

> Alternatively, ClientConn() could return an error when it has a TLS config while called with an empty server name.

I think that this is what happens in practice currently as the TLS handshake will fail as the Server Name will not match the certificate's.

kaworu replied:
Let's follow up in #13087

gandro reviewed (commented):
Overall looks good to me! Some smaller nits. I can do another pass once the base PR has been merged.

sayboras commented on pkg/hubble/relay/server/option.go:
```diff
@@ -15,21 +15,40 @@
+	// ClientTLSMinVersion defines the minimum TLS version that is acceptable
+	// when establishing a connection to a peer. It is not used by
+	// WithClientTLS().
+	ClientTLSMinVersion = tls.VersionTLS13
```
do you feel TLS 1.3 is a little bit strict for min version ?

rolinh replied:
I think that for the connection to Hubble server instances, we can be pretty restrictive in what is allowed as we have full control over both servers (Hubble server instances) and clients (Hubble Relay instances). Hubble server instances are not meant to be accessed other than through Hubble Relay.

sayboras commented on pkg/hubble/relay/server/server.go:
```diff
@@ -33,6 +34,16 @@ import (
+	ErrNoTransportCredentials = errors.New("no transport credentials set (use WithInsecureServer() or set credentials)")
+
+	// ErrNoClientTLSConfig is returned when no client TLS config is set
+	// unlesss WithInsecureClient() is provided.
```
nit: unlesss -> unless

sayboras commented on hubble-relay/cmd/serve/serve.go:
```diff
@@ -117,9 +178,77 @@ func runServe(f flags) error {
+				return nil, fmt.Errorf("cannot load cert '%s': %s", path, err)
+			}
+			if ok := tlsHubbleServerCAs.AppendCertsFromPEM(certPEM); !ok {
+				return nil, fmt.Errorf("cannot process cert '%s': %s", path, err)
```
mostly likely the errr here is nil right ?

rolinh replied:
Indeed, good catch!

sayboras commented on hubble-relay/cmd/serve/serve.go:
```diff
@@ -117,9 +178,77 @@ func runServe(f flags) error {
+		for _, path := range tlsHubbleServerCAPaths {
+			certPEM, err := ioutil.ReadFile(path)
+			if err != nil {
+				return nil, fmt.Errorf("cannot load cert '%s': %s", path, err)
```
%s -> %w if you want

rolinh replied:
I don't think it makes sense to wrap the error here as there is nothing much the caller could do but I can give it more thoughts.

sayboras commented on install/kubernetes/cilium/charts/hubble-tls/templates/secret.yaml:
```diff
@@ -45,15 +45,31 @@ data:
 kind: Secret
 metadata:
-  name: hubble-relay-certs
+  name: hubble-relay-client-certs
```
One question for my understanding, seems like any time helm (e.g. upgarde with re-use values) will generate the new values for certs, will hubble-relay get notified and restart ?

rolinh replied:
This is a good point. I don't think Hubble Relay will get notified. The best approach to address this is probably to watch for cert files changes (typically via fsnotify) and reload the certs when this happens.
There shouldn't be any reason to restart pods when new certs are generated as there is no point breaking existing connections (it's not a problem if certs are re-generated while connections are active because the TLS certs are only useful for the TLS handshake during which 4 session keys are exchanged to encrypt the communication).
I plan to work on this in a follow-up PR. This one is already pretty big.

rolinh committed:
doc: add note about Hubble Relay service disruption during upgrade

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

sayboras reviewed (approved):
LGTM :100:

kaworu commented on hubble-relay/cmd/serve/serve.go:
```diff
@@ -117,9 +178,77 @@ func runServe(f flags) error {
+
+	var tlsHubbleServerCAs *x509.CertPool
+	tlsHubbleServerCAPaths := vp.GetStringSlice(keyTLSHubbleServerCAFiles)
+	switch {
```
any reason to use a `switch` for only two cases (i.e. instead of `if`/`else`)?

kaworu commented on pkg/hubble/peer/types/peer_test.go:
```diff
@@ -149,6 +149,39 @@ func TestFromChangeNotification(t *testing.T) {
+				Name:    "name",
+				Address: "192.0.2.1:4000",
+				Tls: &peerpb.TLS{
+					ServerName: "*.hubble-grpc.cilium.io",
```
nit: this should rather look like `name.default.hubble-grpc.cilium.io` correct?

rolinh replied:
Indeed. Will fix.

kaworu commented on pkg/hubble/peer/types/peer_test.go:
```diff
@@ -149,6 +149,39 @@ func TestFromChangeNotification(t *testing.T) {
+					Port: 4000,
+				},
+				TLSEnabled:    true,
+				TLSServerName: "*.hubble-grpc.cilium.io",
```
ditto

kaworu commented on pkg/hubble/relay/pool/client.go:
```diff
@@ -31,11 +33,27 @@ type GRPCClientConnBuilder struct {
+	opts := make([]grpc.DialOption, len(b.Options))
+	copy(opts, b.Options)
+
+	switch b.TLSConfig {
```
nit: `if`/`else`?

kaworu commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -467,6 +467,79 @@ func TestPeerManager(t *testing.T) {
+									Address: "192.0.1.1",
+									Type:    peerpb.ChangeNotificationType_PEER_ADDED,
+									Tls: &peerpb.TLS{
+										ServerName: "*.hubble-grpc.cilium.io",
```
nit:
```suggestion
										ServerName: "one.default.hubble-grpc.cilium.io",
```

kaworu commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -467,6 +467,79 @@ func TestPeerManager(t *testing.T) {
+									Address: "192.0.1.2",
+									Type:    peerpb.ChangeNotificationType_PEER_ADDED,
+									Tls: &peerpb.TLS{
+										ServerName: "*.hubble-grpc.cilium.io",
```
nit:
```suggestion
										ServerName: "two.default.hubble-grpc.cilium.io",
```

kaworu commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -467,6 +467,79 @@ func TestPeerManager(t *testing.T) {
+									Address: "192.0.1.1",
+									Type:    peerpb.ChangeNotificationType_PEER_DELETED,
+									Tls: &peerpb.TLS{
+										ServerName: "*.hubble-grpc.cilium.io",
```
nit:
```suggestion
										ServerName: "one.default.hubble-grpc.cilium.io",
```

kaworu commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -467,6 +467,79 @@ func TestPeerManager(t *testing.T) {
+								Port: defaults.ServerPort,
+							},
+							TLSEnabled:    true,
+							TLSServerName: "*.hubble-grpc.cilium.io",
```
```suggestion
							TLSServerName: "two.default.hubble-grpc.cilium.io",
```

kaworu commented on pkg/hubble/relay/server/option.go:
```diff
@@ -125,3 +144,81 @@ func WithErrorAggregationWindow(d time.Duration) Option {
+}
+
+// WithClientTLS sets the transport credentials for connecting to peers based
+// on the provided TLS configuration>
```
```suggestion
// on the provided TLS configuration.
```

kaworu commented on pkg/hubble/relay/server/option.go:
```diff
@@ -125,3 +144,81 @@ func WithErrorAggregationWindow(d time.Duration) Option {
+}
+
+// WithClientMTLSFromCert constructs mutual TLS (mTLS) credentials from ca and
+// cert. The certificate is presented to the peer and the provided root
+// certificate authority is used to validate connections to peers.
```
```suggestion
// cert. The certificate is presented to the peers, and the provided root
// certificate authority is used to validate certificates presented by the peers.
```

kaworu commented on pkg/hubble/relay/server/option.go:
```diff
@@ -125,3 +144,81 @@ func WithErrorAggregationWindow(d time.Duration) Option {
+}
+
+// WithInsecureClient disables transport security for connection to Hubble
+// server instances. Transport security is required to WithInsecureClient is
```
```suggestion
// server instances. Transport security is required unless WithInsecureClient is
```

kaworu reviewed (approved):
Awesome PR! Only have some nits.

rolinh committed:
helm: enable mTLS between Hubble server and Hubble Relay intances

Both Hubble server and Hubble Relay now have support for mTLS so it can
be enabled by default.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

aanm commented on install/kubernetes/cilium/charts/hubble-relay/templates/configmap.yaml:
```diff
@@ -0,0 +1,25 @@
+  name: hubble-relay-config
+  namespace: {{ .Release.Namespace }}
+data:
+  config.yaml: |
+    peer-service: unix://{{ .Values.global.hubble.socketPath }}
+    listen-address: {{ .Values.listenHost }}:{{ .Values.listenPort }}
+{{- if .Values.global.debug.enabled }}
+    debug: true
+{{- end }}
+{{- if .Values.dialTimeout }}
+    dial-timeout: {{ .Values.dialTimeout }}
+{{- end }}
+{{- if .Values.retryTimeout }}
+    retry-timeout: {{ .Values.retryTimeout }}
+{{- end }}
+{{- if .Values.sortBufferLenMax }}
+    sort-buffer-len-max: {{ .Values.sortBufferLenMax }}
+{{- end }}
+{{- if .Values.sortBufferDrainTimeout }}
+    sort-buffer-drain-timeout: {{ .Values.sortBufferDrainTimeout }}
+{{- end }}
```
We should have the same implementation when setting up values in the config map as done in Cilium for consistency.
